nixos/modules/services/continuous-integration/hercules-ci-agent/common.nix

```diff
@@ -0,0 +1,211 @@
+        echo "Checking in-memory pathInfoCache expiry"
+        if ! grep 'struct PathInfoCacheValue' src/libstore/store-api.hh >/dev/null; then
+          cat 1>&2 <<EOF
+
+          You are deploying Hercules CI Agent on a system with an incompatible
+          nix-daemon. Please
+           - either upgrade Nix to version 2.4.0 (when released),
+           - or set option services.hercules-ci-agent.patchNix = true;
+           - or set option nix.package to a build of Nix 2.3 with this patch applied:
+               https://github.com/NixOS/nix/pull/3405
+
+          The patch is required for Nix-daemon clients that expect a change in binary
+          cache contents while running, like the agent's evaluator. Without it, import
+          from derivation will fail if your cluster has more than one machine.
+          We are conservative with changes to the overall system, which is why we
+          keep changes to a minimum and why we ask for confirmation in the form of
+          services.hercules-ci-agent.patchNix = true before applying.
+
+        EOF
+          exit 1
+        fi
+      '';
+      installPhase = "touch $out";
+    };
+
+  patchedNix = lib.mkIf (!lib.versionAtLeast pkgs.nix.version "2.4.0") (
+    if lib.versionAtLeast pkgs.nix.version "2.4pre"
+    then lib.warn "Hercules CI Agent module will not patch 2.4 pre-release. Make sure it includes (equivalently) PR #3043, commit d048577909 or is no older than 2020-03-13." pkgs.nix
+    else pkgs.nix.overrideAttrs (
+      o: {
+        patches = (o.patches or []) ++ [ ./issue-3398-path-info-cache-ttls-backport-2.3.patch ];
+      }
+    )
+  );
+in
+{
+  imports = [
+    ./gc.nix
+  ];
+
+  options.services.hercules-ci-agent = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Enable to run Hercules CI Agent as a system service.
+
+        <link xlink:href="https://hercules-ci.com">Hercules CI</link> is a
+        continuous integation service that is centered around Nix.
+
+        Support is available at <link xlink:href="mailto:help@hercules-ci.com">help@hercules-ci.com</link>.
+      '';
+    };
+    patchNix = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to set nix.package to a patched pkgs.nix if necessary.
+
+        We strive for minimal patching. This option is only for your convenience,
+        when an important patch is not accepted into the stable branch for
+        example.
+      '';
+    };
+    checkNix = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        Whether to make sure that the system's Nix (nix-daemon) is compatible.
+
+        If you set this to false, please keep up with the change log.
+      '';
+    };
+    baseDirectory = mkOption {
+      type = types.path;
+      default = "/var/lib/hercules-ci-agent";
+      description = "State directory (secrets, work directory, etc) for agent";
+    };
```

Comment:
Can you use `StateDirectory=` in the unit?

Reply:
This maps to an entry in the config file. I suppose I could derive `StateDirectory` from it, but I'm not sure what would be the benefit of doing so.

Reply:
Hmm, I wasn't specific here - if this were using `DynamicUser` (which it can't, due to suid support), We could just have defaulted the state directory to this location.

It's probably still beneficial to distinguish between the home directory of the user running the service (which could be somewhere else than `/var/lib/buildkite-agent`), and state files (which should always live in `/var/lib/$serviceName`.

Reply:
It currently depends on dotfiles to configure automatic S3 push for example, so the default home `/var/empty` wouldn't work. Disallowed or transient dotfiles are normally the benefit I suppose. Is there another benefit that could offset the migration that will be required for current installations?

Reply:
Relevant issue: https://github.com/hercules-ci/hercules-ci-agent/issues/239
For now I don't see a reason to use `StateDirectory`.